### DISCUSSION TITLE
Has Anyone Used a ChromeBook for MicroPython Development?

### DISCUSSION THREAD

**Discussion by DavesCodeMusings** [micropython/micropython#15123] [General]
I'm undertaking a project to corrupt the minds of middle-school students with the wonders of microcontrollers and MicroPython. I assumed the school would have some sort of lab with PCs or Macs capable of running VS Code and MPRemote or even Thonny. I was wrong. It's ChromeBooks as far as the eye can see.

Has anyone successfully used a ChromeBook to develop with MicroPython?

There's a lot listed under "Not Supported Yet" on this ChromeBook Help KB article: https://support.google.com/chromebook/answer/9145439?hl=en

This includes many USB devices, so I have a strong suspicion I'll be looking at my "plan B". But ,I thought I would check with the group first to see if anyone has made this work.

**Comment by Josverl** [comment]
Chromebook should support webusb, and tools like makecode as well.

So while I've not used the combo before, I think that centering your search around tools that allow flashing or transferring code over webusb will deliver results

**Reply by DavesCodeMusings** [reply]
I did some searching around makecode and ESP32 (the microcontroller I was planning to use) and most of what turned up was along the lines of "Hey, is ESP32 supported yet?" Some of the posts were pretty old, so I'll keep digging. Thanks for the suggestion.

**Comment by victorallume** [comment]
I've done it with Micro:bit devices. They're more expensive than ESP32, but there's a much larger education community and support around them, so that can make it worthwhile. Is there any particular reason you're leaning towards ESP32?

**Reply by DavesCodeMusings** [reply]
I'm looking at ESP32 because of its WiFi capabilities. This makes it great for Internet of Things (IoT) type experiments. So far I've done a lab centered around a [ESP8266-based DHT22 temperature/humidity sensor with a REST API](https://github.com/DavesCodeMusings/thimble/tree/main/examples/esp8266-dht22)

In this scenario, the students were using their Chromebook's web browser as a REST client to read values from the device via its API. I'd like to expand on this and let students start hacking the device to eventually do their own IoT projects. Ideally, they would each have their own device, so the price point of an ESP32C is hard to beat ($8 US).

**Reply by victorallume** [reply]
The micro:bit does have bluetooth giving you some IoT capability. Although micropython's then simplifies this to a `radio` module, which is _much_ simpler to understand than BLE, but at the same time makes it not interoperable with other bluetooth devices (last I checked anyway - only lets micro:bit -> micro:bit comms).

Yes, the price of the ESP32 makes it very tempting, but it's one of those things that might take a lot more time to get running, and in some places, time is money ;)

The other possibility might be to use webrepl. Once micropython is flashed onto the device (you'd need to do this on a non-chromebook), you can connect and run repl commands and send files. Though I can imagine this might end up being confusing (and congesting the wifi network) if you have 20 students all doing this at once - it would be worth testing this out with a handful of devices at once to pre-empt problems with a whole class-full.

One other half-formed idea: serial-over-bluetooth. It would again require flashing (probably a custom build), and again might have issues when it comes to pairing when you have 20 devices all next to each other.

**Comment by mattytrentini** [comment]
So a few thoughts maybe worth sharing!

There are a few projects that have overcome the general issues with ChromeOS and MicroPython, but this area is a little _immature_.

There are perhaps four aspects to discuss:

1. Deploying MicroPython firmware to a device
2. Interacting with a live REPL
3. Copying files to/from the device filesystem
4. Having a usable development environment

Support for 1 depends on the microcontroller and bootloader. Those that use `UF2` (rp2, samd, some ESP32's - but potentially others. mimxrt uses something similar?) should work _now_. It's just a matter of putting a micro in "UF2 mode" so that it appears as a USB mass storage device and then copying files to it. However, other devices that use `DFU` (stm32) or `esptool` (ESP32) are a _little_ more challenging - but possible! There are examples of DFU ([WebDFU](https://devanlai.github.io/projects/webdfu/)) and esptool ([esptool-js](https://github.com/espressif/esptool-js/, [BIPES](https://bipes.net.br/flash/esp-web-tools/)), working in the browser over WebUSB/WebSerial. These work at least as proof-of-concepts and could _potentially_ be integrated into the [MicroPython Downloader](https://micropython.org/download/) so they could be deployed directly from that site.

Number 2 requires supporting a serial connection from the browser. Again, there are good proof-of-concepts ([BIPES web-serial-terminal](https://github.com/rafaelaroca/web-serial-terminal), [serialTerminal](https://www.serialterminal.com/)) and they wouldn't require too much polish to be very usable. 

Item 3 would need to extend 2 to add `mpremote`-style file transfer support. Shouldn't be a huge leap but I'm not aware of any efforts to do so.

4 is about having a development IDE in the browser - something like [VS Code for the Web](https://code.visualstudio.com/docs/editor/vscode-web).

As I said, a lot of this is available now, it's just not a great experience. I like to imagine [github.dev](https://docs.github.com/en/codespaces/the-githubdev-web-based-editor) (basically VS Code for the Web hosted on Github) with a [web extension](https://code.visualstudio.com/api/extension-guides/web-extensions) to interact with a connected device - and it doesn't seem terribly difficult!

**Reply by DavesCodeMusings** [reply]
Of the research I've done so far, it seems like the [limitations of USB in ChomeOS Linux mode](https://support.google.com/chromebook/answer/9145439?hl=en) might be the biggest hurdle. (See the bottom of the KB article _Check what's not supported yet_.) This would impact aspects 1, 2, and 3 that you listed. (Had I known about this from the start, I may not have started down the Chromebook path.)

As I go, I'm also finding out about the Chromebook administrative policies that are currently preventing access to Linux mode and the internet policies that are restricting sites like vscode.dev and github.com. The internet restrictions have the potential to impact aspect 4.

In the end, I may throw up my hands and say "Raspberry Pi OS development environment it is, then!"

I'm tracking my progress on a [GitHub kanban board](https://github.com/users/DavesCodeMusings/projects/2) if anyone is interested in following along. Once I determine a direction (Chromebook, Raspberry Pi OS on ARM, or Windows/Linux on x86), I'll post something in Show & Tell about it. My goal is an inexpensive option that the students can build and maintain with minimal help from teachers and volunteers.

**Reply by gkhan1234** [reply]
I'm going through the same issue here. I'm trying to build a server on the ESP32 to allow a direct connection for hacking. Any browser should work.

**Reply by DavesCodeMusings** [reply]
> I'm going through the same issue here. I'm trying to build a server on the ESP32 to allow a direct connection for hacking. Any browser should work.

In the end, I created a Docker container to run web-based VS Code, MPRemote, ESPTool, etc. on an old Raspberry Pi. https://github.com/DavesCodeMusings/school-coding-lab/tree/main/rpi/code-server

**Comment by DavesCodeMusings** [comment]
I'm quickly coming to the conclusion of: _It might be possible, but it's gonna be painful._ 

Add in the administrative policies governing what internet sites are available for these devices and the outlook is even more dismal.

So, I'm pivoting and looking at other solutions for student developer workstations. Currently, that's looking like upcycling discarded Raspberry Pis into something useful and making MicroPython firmware and other files available on the Pis via USB flash drive sneaker net.

I appreciate everyone who's chimed in with suggestions.

If anybody is interested in the Raspberry Pi approach, I've started a [GitHub project repo](https://github.com/DavesCodeMusings/school-coding-lab/) for it. It's full of a lot of TODO placeholders right now, but I've already got a line on some donated Raspberry Pi hardware. My hope is to have something ready for a small pilot group of students by autumn.

**Comment by urfdvw** [comment]
Not sure if this thread is still active or not but in 2025, Viper IDE is good enough for this purpose.
